File: sdks/js/tests/session.test.ts
import assert from "node:assert/strict";
import { afterEach, beforeEach, describe, it } from "node:test";

import { Client, Session, Socket, Tools } from "../src/index.js";
import { TestRouter, type Connection } from "./router.js";

describe("Socket", () => {
  let router: TestRouter;
  let api: Client;

  beforeEach(async () => {
    router = await TestRouter.start();
    api = new Client({ url: router.url, customerId: "local" });
  });

  afterEach(async () => {
    await router.stop();
  });

  it("yields the frames the router sends, in order", async () => {
    const opening = Socket.open(api.backend, await api.backend.socketURL("/v1/stt/stream"));
    const connection = await router.socket();
    const socket = await opening;

    connection.send({ type: "transcript", text: "one" });
    connection.send({ type: "transcript", text: "two" });

    const read: unknown[] = [];
    for await (const message of socket.messages()) {
      read.push(message);
      if (read.length === 2) {
        break;
      }
    }

    assert.deepEqual(read, [
      { type: "transcript", text: "one" },
      { type: "transcript", text: "two" },
    ]);
    socket.close();
  });

  it("hands audio back as bytes rather than as a frame", async () => {
    const opening = Socket.open(api.backend, await api.backend.socketURL("/v1/tts/stream"));
    const connection = await router.socket();
    const socket = await opening;

    connection.socket.send(Buffer.from([1, 2, 3]));

    for await (const message of socket.messages()) {
      assert.ok(message instanceof Uint8Array);
      assert.deepEqual([...message], [1, 2, 3]);
      break;
    }
    socket.close();
  });

  it("ends the stream when the router closes it", async () => {
    const opening = Socket.open(api.backend, await api.backend.socketURL("/v1/llm/stream"));
    const connection = await router.socket();
    const socket = await opening;

    connection.socket.close();

    const read: unknown[] = [];
    for await (const message of socket.messages()) {
      read.push(message);
    }
    assert.deepEqual(read, []);
    assert.equal(socket.open, false);
  });

  it("refuses a send on a socket that is no longer open", async () => {
    const opening = Socket.open(api.backend, await api.backend.socketURL("/v1/llm/stream"));
    await router.socket();
    const socket = await opening;
    socket.close();

    assert.throws(() => socket.send({ type: "respond" }));
  });

  it("reports an upgrade the router refused rather than resolving", async () => {
    await assert.rejects(() =>
      Socket.open(api.backend, "ws://127.0.0.1:1/v1/agents/sessions/x/events"),
    );
  });
});

describe("Session", () => {
  let router: TestRouter;
  let api: Client;

  beforeEach(async () => {
    router = await TestRouter.start();
    api = new Client({ url: router.url, customerId: "local" });
    router.serve("POST", "/v1/agents/sessions", {
      status: 201,
      body: {
        id: "sess_1",
        call_id: "demo",
        call_type: "agent",
        user_id: "john",
        agent_id: "john",
        state: "live",
        created_at: "2026-01-01T00:00:00Z",
      },
    });
  });

  afterEach(async () => {
    await router.stop();
  });

  /** Opens a session and hands back both ends of it. */
  async function opened(tools?: Tools): Promise<[Session, Connection]> {
    const opening = Session.open(api, { call_id: "demo" }, tools ? { tools } : {});
    const connection = await router.socket();
    return [await opening, connection];
  }

  it("creates the session and watches the socket for it", async () => {
    const [session, connection] = await opened();

    assert.equal(session.id, "sess_1");
    assert.equal(connection.path, "/v1/agents/sessions/sess_1/events");
    assert.equal(connection.query.get("customer_id"), "local");
    await session.close();
  });

  it("asks for interim speech only when told to", async () => {
    const opening = Session.open(api, { call_id: "demo" }, { interim: true });
    const connection = await router.socket();
    const session = await opening;

    assert.equal(connection.query.get("interim"), "true");
    await session.close();
  });

  it("declares the registered tools on the session it creates", async () => {
    const tools = new Tools().register({
      name: "get_weather",
      description: "Get the weather somewhere",
      parameters: { type: "object", properties: { city: { type: "string" } } },
      run: () => "sunny",
    });

    const [session] = await opened(tools);

    const body = router.last.body as { tools: { name: string; description: string }[] };
    assert.equal(body.tools.length, 1);
    assert.equal(body.tools[0]?.name, "get_weather");
    await session.close();
  });

  it("closes the session in the backend when the socket cannot be watched", async () => {
    const unwatchable = new Client({ url: router.url, customerId: "local" });
    router.serve("DELETE", "/v1/agents/sessions/sess_1", { status: 204 });
    await router.stop();

    await assert.rejects(() => Session.open(unwatchable, { call_id: "demo" }));
  });

  it("hands the backend's events to whoever is reading them", async () => {
    const [session, connection] = await opened();

    connection.send({ type: "joined" });
    connection.send({
      type: "heard",
      text: "hello",
      participant: { id: "p1", user_id: "ana", name: "Ana" },
    });

    const read = [];
    for await (const event of session.events()) {
      read.push(event);
      if (read.length === 2) {
        break;
      }
    }

    assert.equal(read[0]?.kind, "joined");
    assert.equal(read[1]?.text, "hello");
    assert.deepEqual(read[1]?.participant, { id: "p1", userId: "ana", name: "Ana" });
    await session.close();
  });

  it("runs a tool the model asked for and answers with what it returned", async () => {
    const tools = new Tools().register<{ city: string }>({
      name: "get_weather",
      description: "Get the weather somewhere",
      run: (input) => ({ city: input.city, sky: "clear" }),
    });
    const [session, connection] = await opened(tools);

    connection.send({
      type: "tool_call",
      id: "call_1",
      name: "get_weather",
      arguments: JSON.stringify({ city: "Boulder" }),
    });

    const answer = await connection.next();
    assert.deepEqual(answer, {
      type: "tool_result",
      tool_call_id: "call_1",
      output: JSON.stringify({ city: "Boulder", sky: "clear" }),
    });
    await session.close();
  });

  it("tells the model a tool did not work, since it is mid-sentence waiting", async () => {
    const tools = new Tools().register({
      name: "lookup",
      description: "Look something up",
      run: () => {
        throw new Error("the database is down");
      },
    });
    const [session, connection] = await opened(tools);

    connection.send({ type: "tool_call", id: "call_1", name: "lookup", arguments: "{}" });

    assert.deepEqual(await connection.next(), {
      type: "tool_result",
      tool_call_id: "call_1",
      error: "the database is down",
    });
    await session.close();
  });

  it("reports a tool the model asked for that nothing is registered as", async () => {
    const [session, connection] = await opened(new Tools());

    connection.send({ type: "tool_call", id: "call_1", name: "unknown", arguments: "{}" });

    const answer = await connection.next();
    assert.match(String(answer["error"]), /nothing is registered as unknown/);
    await session.close();
  });

  it("aborts a running tool when the model cancels it", async () => {
    let aborted = false;
    const tools = new Tools().register({
      name: "slow",
      description: "Take a while",
      run: (_input, signal) =>
        new Promise((resolve) => {
          signal.addEventListener("abort", () => {
            aborted = true;
            resolve("abandoned");
          });
        }),
    });
    const [session, connection] = await opened(tools);

    connection.send({ type: "tool_call", id: "call_1", name: "slow", arguments: "{}" });
    await new Promise((resolve) => setTimeout(resolve, 20));
    connection.send({ type: "tool_cancel", id: "call_1" });

    await connection.next();
    assert.equal(aborted, true);
    await session.close();
  });

  it("does not hand a tool call to whoever is reading events", async () => {
    const tools = new Tools().register({
      name: "ping",
      description: "Answer",
      run: () => "pong",
    });
    const [session, connection] = await opened(tools);

    connection.send({ type: "tool_call", id: "call_1", name: "ping", arguments: "{}" });
    connection.send({ type: "responded", text: "done" });

    for await (const event of session.events()) {
      assert.equal(event.kind, "responded");
      break;
    }
    await session.close();
  });

  it("sends what the caller asked the conversation to do", async () => {
    const [session, connection] = await opened();

    session.say("one moment");
    assert.deepEqual(await connection.next(), { type: "say", text: "one moment" });

    session.respond("what is the weather", { interrupt: true });
    assert.deepEqual(await connection.next(), { type: "interrupt" });
    assert.deepEqual(await connection.next(), {
      type: "respond",
      text: "what is the weather",
    });

    session.setInstructions("be brief");
    assert.deepEqual(await connection.next(), {
      type: "instructions",
      instructions: "be brief",
    });

    await session.close();
  });

  it("ends the conversation over the socket it is being watched on", async () => {
    const [session, connection] = await opened();

    const closing = session.close();
    assert.deepEqual(await connection.next(), { type: "close" });
    await closing;

    assert.equal(session.live, false);
  });

  it("stops waiting when the conversation ends", async () => {
    const [session, connection] = await opened();

    connection.socket.close();

    await session.wait();
    assert.equal(session.live, false);
  });
});

File: sdks/dart/core/test/session_test.dart
import 'dart:async';

import 'package:test/test.dart';
import 'package:vision_agents_core/vision_agents_core.dart';

import 'support/fixtures.dart';
import 'support/router.dart';

void main() {
  late TestRouter router;
  late VisionAgents agents;

  setUp(() async {
    router = await TestRouter.start();
    router.answer('POST /v1/agents/sessions', Answer(201, sessionJson()));
    agents = VisionAgents(url: router.url, customerId: 'acme');
  });

  tearDown(() async {
    agents.close();
    await router.close();
  });

  group('AgentSession', () {
    test('opens the socket as a device, with decisions off, before it returns', () async {
      final chat = await agents.withUser('ada').chat();

      expect(chat.isConnected, isTrue);
      final upgrade = router.upgrades.single;
      expect(upgrade.path, '/v1/agents/sessions/s1/events');
      expect(upgrade.query, {'decisions': 'false', 'customer_id': 'acme', 'user_id': 'ada'});
      expect(upgrade.header('Stream-Auth-Type'), isNull);
      await chat.close();
    });

    test('shows what was typed at once and sends it as a respond', () async {
      final chat = await agents.chat();
      final socket = await router.socket();

      chat.send('  What are your hours?  ');

      expect(chat.turns.single.text, 'What are your hours?');
      expect(await socket.next(), {'type': 'respond', 'text': 'What are your hours?'});
      await chat.close();
    });

    test('folds the router frames into the transcript and the state', () async {
      final chat = await agents.chat();
      final socket = await router.socket();

      socket
        ..send('{"type":"responding","turn_id":"t1","participant":{},"prompt":"hi"}')
        ..send('{"type":"response_delta","turn_id":"t1","text":"Nine"}')
        ..send(
          '{"pending_work":false,"type":"responded","turn_id":"t1","text":"Nine to five.","time_to_first_token_ms":90}',
        );

      await until(() => chat.turns.isNotEmpty && chat.state == ConversationState.idle);
      expect(chat.turns.single.text, 'Nine to five.');
      await chat.close();
    });

    test('a widget building from the conversation sees each change as it lands', () async {
      final chat = await agents.chat();
      final socket = await router.socket();
      final states = <ConversationState>[];
      final subscription = chat.conversation.stream.listen((c) => states.add(c.state));

      socket.send('{"type":"responding","turn_id":"t1"}');
      socket.send('{"type":"responded","turn_id":"t1","text":"Hi."}');

      await until(() => states.length >= 3);
      expect(states, [
        ConversationState.idle,
        ConversationState.responding,
        ConversationState.idle,
      ]);
      await subscription.cancel();
      await chat.close();
    });

    test('skips a frame it cannot read rather than ending the conversation', () async {
      final chat = await agents.chat();
      final socket = await router.socket();

      socket
        ..send('not json')
        ..send('{"no":"type"}')
        ..send('{"type":"response_delta","turn_id":"t1","text":"still here"}');

      await until(() => chat.turns.isNotEmpty);
      expect(chat.turns.single.text, 'still here');
      expect(chat.isConnected, isTrue);
      await chat.close();
    });

    test('runs a tool on this side and answers the model with what it returned', () async {
      final asked = <Object?>[];
      final chat = await agents.chat(
        SessionOptions(
          tools: [
            AgentTool(
              name: 'lookup_order',
              description: 'Look up an order.',
              run: (arguments) async {
                asked.add(arguments['order_id']);
                return 'Order A-1042: 2 wool throws.';
              },
            ),
          ],
        ),
      );
      final socket = await router.socket();

      socket.send(
        '{"type":"tool_call","id":"c1","name":"lookup_order",'
        r'"arguments":"{\"order_id\":\"A-1042\"}","command_id":"","turn_id":"t1"}',
      );

      expect(await socket.next(), {
        'type': 'tool_result',
        'tool_call_id': 'c1',
        'output': 'Order A-1042: 2 wool throws.',
        'error': '',
      });
      expect(asked, ['A-1042']);
      await chat.close();
    });

    test('tells the model a tool did not work, since it is mid-sentence waiting', () async {
      final chat = await agents.chat(
        SessionOptions(
          tools: [
            AgentTool(
              name: 'lookup_order',
              description: 'Look up an order.',
              run: (_) async => throw StateError('the order service is down'),
            ),
          ],
        ),
      );
      final socket = await router.socket();

      socket.send('{"type":"tool_call","id":"c1","name":"lookup_order","arguments":"{}"}');

      final result = await socket.next();
      expect(result['output'], '');
      expect(result['error'], contains('the order service is down'));
      await chat.close();
    });

    test("repeats a durable command's ids, so no other command can adopt the result", () async {
      final chat = await agents.chat(
        SessionOptions(
          tools: [AgentTool(name: 'now', description: 'The time.', run: (_) async => 'noon')],
        ),
      );
      final socket = await router.socket();

      socket.send(
        '{"type":"tool_call","id":"c1","name":"now","arguments":"{}","command_id":"cmd-1","turn_id":"t1"}',
      );

      expect(await socket.next(), {
        'type': 'tool_result',
        'tool_call_id': 'c1',
        'output': 'noon',
        'error': '',
        'command_id': 'cmd-1',
        'turn_id': 't1',
      });
      await chat.close();
    });

    test('leaves a tool it does not have for whoever does, rather than answering first', () async {
      final chat = await agents.chat(
        SessionOptions(
          tools: [AgentTool(name: 'mine', description: 'Mine.', run: (_) async => 'mine')],
        ),
      );
      final socket = await router.socket();

      socket
        ..send('{"type":"tool_call","id":"c1","name":"somebody_elses","arguments":"{}"}')
        ..send('{"type":"tool_call","id":"c2","name":"mine","arguments":"{}"}');

      expect((await socket.next())['tool_call_id'], 'c2');
      await chat.close();
    });

    test('a tool call the router cancelled is not answered', () async {
      final release = Completer<String>();
      final chat = await agents.chat(
        SessionOptions(
          tools: [
            AgentTool(name: 'slow', description: 'Slow.', run: (_) => release.future),
            AgentTool(name: 'fast', description: 'Fast.', run: (_) async => 'fast'),
          ],
        ),
      );
      final socket = await router.socket();
      final cancelled = chat.events().firstWhere(
        (event) => event.kind == AgentEventKind.toolCancel,
      );

      socket
        ..send('{"type":"tool_call","id":"c1","name":"slow","arguments":"{}"}')
        ..send('{"type":"tool_cancel","id":"c1","command_id":"","turn_id":""}');
      await cancelled;
      release.complete('too late');
      socket.send('{"type":"tool_call","id":"c2","name":"fast","arguments":"{}"}');

      expect((await socket.next())['tool_call_id'], 'c2');
      await chat.close();
    });

    test('answering a tool does not depend on anybody reading events', () async {
      final chat = await agents.chat(
        SessionOptions(
          tools: [AgentTool(name: 'now', description: 'The time.', run: (_) async => 'noon')],
        ),
      );
      final socket = await router.socket();

      socket.send('{"type":"tool_call","id":"c1","name":"now","arguments":"{}"}');

      expect((await socket.next())['output'], 'noon');
      await chat.close();
    });

    test('two readers of events each get every event', () async {
      final chat = await agents.chat();
      final socket = await router.socket();
      final first = chat.events().map((event) => event.type).take(2).toList();
      final second = chat.events().map((event) => event.type).take(2).toList();

      socket
        ..send('{"type":"joined","at":"2026-09-24T15:54:56Z"}')
        ..send('{"type":"astonished"}');

      expect(await first, ['joined', 'astonished']);
      expect(await second, ['joined', 'astonished']);
      await chat.close();
    });

    test('a reader that falls far behind is finished rather than buffered forever', () async {
      final chat = await agents.chat();
      final socket = await router.socket();
      final slow = chat.events();

      for (var i = 0; i < 300; i++) {
        socket.send('{"type":"response_delta","turn_id":"t1","text":"."}');
      }
      await until(() => chat.turns.isNotEmpty && chat.turns.single.text.length == 300);

      expect(await slow.length, 256);
      expect(chat.isConnected, isTrue);
      await chat.close();
    });

    test('a session the router ended is over, with no failure to report', () async {
      final chat = await agents.chat();
      final socket = await router.socket();

      socket.send('{"type":"left","at":"2026-09-24T15:54:56Z"}');
      await socket.close(1000, 'the session ended');

      await until(() => chat.connection.value is Disconnected);
      expect(chat.failure, isNull);
      expect(chat.state, ConversationState.ended);
    });

    test('a socket lost mid-conversation reports why, with its close code', () async {
      final chat = await agents.chat();
      final socket = await router.socket();

      await socket.close(4000, 'draining');

      await until(() => chat.connection.value is Disconnected);
      expect(
        chat.failure,
        isA<SocketClosedException>()
            .having((e) => e.code, 'code', 4000)
            .having((e) => e.reason, 'reason', 'draining'),
      );
      expect(() => chat.send('anyone?'), throwsA(isA<SocketClosedException>()));
    });

    test('closing says so over the socket and is safe to repeat', () async {
      final chat = await agents.chat();
      final socket = await router.socket();

      await chat.close();
      await chat.close();

      expect(await socket.next(), {'type': 'close'});
      await socket.done;
      expect(socket.closeCode, 1000);
      expect(chat.connection.value, const Disconnected());
    });

    test(
      'a session whose socket cannot open is ended rather than left holding the agent',
      () async {
        router
          ..answer('DELETE /v1/agents/sessions/s1', const Answer(204))
          ..refuseSockets = true;

        await expectLater(agents.chat(), throwsA(isA<TransportException>()));
        expect(router.last('DELETE /v1/agents/sessions/s1').method, 'DELETE');
      },
    );

    test('a fork keeps the tools, since they are here in this process', () async {
      router.answer('POST /v1/agents/sessions/s1/fork', Answer(201, sessionJson(id: 's2')));
      final chat = await agents.chat(
        SessionOptions(
          tools: [AgentTool(name: 'now', description: 'The time.', run: (_) async => 'noon')],
        ),
      );

      final fork = await chat.fork(const ForkOptions(responseId: 'r1'));
      await fork.start();
      final socket = await router.socket(1);
      socket.send('{"type":"tool_call","id":"c1","name":"now","arguments":"{}"}');

      expect(fork.id, 's2');
      expect((await socket.next())['output'], 'noon');
      await fork.close();
      await chat.close();
    });
  });
}
